### PR TITLE
feat(provision): add support for thin provision lvm volumes

### PR DESCRIPTION
**Pull request by prateekpandey14**
Signed-off-by: prateekpandey14 <prateekpandey14@gmail.com>

## Pull Request template

**What this PR does?**:
Adds support for thin provision lvm volumes. To create a thin provisioned volumes,  following tasks needs to be performed:

- Create a volume group with the vgcreate command.
- Create a thin pool with the lvcreate command.
- Create a thin volume in the thin pool with the lvcreate command.

**Does this PR require any upgrade changes?**:
No

**If the changes in this PR are manually verified, list down the scenarios covered:**:
- Thin Provision needs to be enable via StorageClass, for example
```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: openebs-lvmsc
allowVolumeExpansion: true
parameters:
  volgroup: "lvmvg"
  thinprovision: "yes"
provisioner: local.csi.openebs.io
```

- Created one thin provision and normal volume to test the changes

```bash
$ lvs
  LV                                       VG    Attr       LSize Pool           Origin Data%  Meta%  Move Log Cpy%Sync Convert
  lvmvg_thinpool                           lvmvg twi-aotz-- 5.00g                       2.88   11.57                           
  pvc-b8f44bb4-6c6c-4c8f-829e-0a1e4af031e1 lvmvg -wi-ao---- 5.00g                                                              
  pvc-e5162004-f857-4104-b051-62c5ba24360b lvmvg Vwi-aotz-- 5.00g lvmvg_thinpool        2.88                                   


```
- Created the volume snapshot  for thin provisioned volume

```bash
$ lvs
  LV                                       VG    Attr       LSize Pool           Origin                                   Data%  Meta%  Move Log Cpy%Sync Convert
  bee6faff-1fd5-4c19-9c3d-aaa34f15c788     lvmvg sri-a-s--- 5.00g                pvc-a5190bee-f48c-4765-815c-fd6838dd2b4c 0.01                                   
  lvmvg_thinpool                           lvmvg twi-aotz-- 5.00g                                                         2.88   11.57                           
  pvc-a5190bee-f48c-4765-815c-fd6838dd2b4c lvmvg owi-aotz-- 5.00g lvmvg_thinpool                                          2.88                                   


```

**Any additional information for your reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [x] Fixes #<issue number>
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated?
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

### CODE REVIEW

File: pkg/lvm/lvm_util.go [hunk 1]
```diff
@@ -86,7 +86,33 @@ func buildLVMCreateArgs(vol *apis.LVMVolume) []string {
 	}
 
 	LVMVolArg = append(LVMVolArg, vol.Spec.VolGroup)
+	return LVMVolArg
+}
+
+// builldThinLVMCreateArgs returns lvcreate command for thin volume
+func buildThinLVMCreateArgs(vol *apis.LVMVolume) []string {
```
Comment (pawanpraka1) [inline pkg/lvm/lvm_util.go:93]: we can merge this with buildLVMCreateArgs itself?

**Review by pawanpraka1 — COMMENTED** [review]
looks good. Do we need to do something for resize in case of thin provisioned volumes?

File: deploy/lvm-operator.yaml [hunk 2]
```diff
@@ -109,6 +109,15 @@ spec:
                 - "yes"
                 - "no"
                 type: string
+              thinProvision:
+                description: ThinProvision specifies whether logical volumes can be
+                  thinly provisioned. If it is not set to "yes", then the LVM LocalPV
+                  Driver will create thinProvision i.e. logical volumes that are larger
+                  than the available extents.
```
Comment (akhilerm) [inline deploy/lvm-operator.yaml:116]: The description is wrong. If `thinProvision` is set to `"yes"`, then then LVM driver will create thin provisioned volumes . .

File: pkg/lvm/lvm_util.go [hunk 3]
```diff
@@ -86,7 +86,33 @@ func buildLVMCreateArgs(vol *apis.LVMVolume) []string {
...
+	var LVMVolArg []string
+
+	volume := vol.Name
+	size := vol.Spec.Capacity + "b"
+	pool := vol.Spec.VolGroup + "_thinpool"
```
Comment (akhilerm) [inline pkg/lvm/lvm_util.go:98]: Does LVM change the name like this?

**Review by akhilerm — COMMENTED** [review]
Some questions
1. What will happen if snapshot of a thin provisioned volume is taken?

File: pkg/lvm/lvm_util.go (hunk 3 above)
Comment (prateekpandey14) [reply pkg/lvm/lvm_util.go:98]: its a thin pool needs to be created for a thin provisioned volume

File: pkg/lvm/lvm_util.go (hunk 1 above)
Comment (prateekpandey14) [reply pkg/lvm/lvm_util.go:93]: i did this earlier in single func, which is kind of confusing with so many if conditions checks, so thought of making it clear using different func

**Comment by prateekpandey14** [comment]
>What will happen if snapshot of a thin provisioned volume is taken?

Normal Snapshot will be created successfully, have added the example output in description.

File: pkg/lvm/lvm_util.go (hunk 1 above)
Comment (pawanpraka1) [reply pkg/lvm/lvm_util.go:93]: @prateekpandey14 we should not create a new function for each argument. It seems likewe need one if condition for thin provisioning to do all the related task there?

File: pkg/lvm/lvm_util.go [hunk 4]
```diff
@@ -86,7 +86,33 @@ func buildLVMCreateArgs(vol *apis.LVMVolume) []string {
...
+
+	if len(vol.Spec.Capacity) != 0 {
+		// check if thin pool exists for given volumegroup
+		if !lvThinExists(vol.Spec.VolGroup, pool) {
```
Comment (pawanpraka1) [inline pkg/lvm/lvm_util.go:102]: why this check? what will happen if thin pool is there?

Comment (prateekpandey14) [reply pkg/lvm/lvm_util.go:102]: if the 
- thin pool exists for given volumegroup: create command will not have pool args( -L flag hrere) hence the pool will not be created. (single thin pool can be used for multiple thin volumes via storageclass)
- thin pool not exists for given volumegroup: create command will have pool args  included,which will create the thin pool as well with thin volume.

File: pkg/lvm/lvm_util.go [hunk 5]
```diff
@@ -86,7 +86,33 @@ func buildLVMCreateArgs(vol *apis.LVMVolume) []string {
...
+			LVMVolArg = append(LVMVolArg, "-L", size)
+		}
+	}
 
+	// command to create thinpool and thin volume
+	// `lvcreate -L 1G -T lvmvg/mythinpool -V 1G -n thinvol`
+	if strings.TrimSpace(vol.Spec.ThinProvision) == "yes" {
```
Comment (prateekpandey14) [inline pkg/lvm/lvm_util.go:109]: this extra check can be removed here since we are checking this already in caller

**Review by pawanpraka1 — APPROVED** [review]
Looks good

**Pull request merged**